aten/src/ATen/native/cuda/AdaptiveMaxPooling3d.cu

```diff
@@ -0,0 +1,474 @@
+    isizeW = input.size(4);
+
+    istrideD = input.stride(1);
+    istrideT = input.stride(2);
+    istrideH = input.stride(3);
+    istrideW = input.stride(4);
+
+    output.resize_({sizeB, sizeD, osizeT, osizeH, osizeW});
+    indices.resize_({sizeB, sizeD, osizeT, osizeH, osizeW});
+
+    totalZ = sizeB * sizeD * osizeT;
+  }
+
+  int64_t offsetZ = 0;
+  dim3 threads(32, 8);
+  // each H*W plane is processed by blocksH thread blocks
+  int blocksH = std::max((int)(16L / totalZ), 1);
+  while (totalZ > 0) {
+    dim3 blocks(totalZ > 65535 ? 65535 : totalZ, blocksH);
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.scalar_type(),
+      "adaptive_max_pool3d_cuda",
+      [&] {
+        scalar_t *input_data = input.data<scalar_t>();
+        scalar_t *output_data = output.data<scalar_t>();
+        int64_t *indices_data = indices.data<int64_t>();
+
+        adaptivemaxpool<<<blocks, threads, 0, at::cuda::getCurrentCUDAStream()>>>(
+          input_data, output_data, indices_data, isizeT, isizeH, isizeW,
+          osizeT, osizeH, osizeW, istrideD, istrideT, istrideH, istrideW, offsetZ
+      );
+
+      totalZ -= 65535;
+      offsetZ += 65535;
+      }
+    );
+    THCudaCheck(cudaGetLastError());
+  }
+}
+
+void adaptive_max_pool3d_backward_out_cuda_template(
+           Tensor& gradInput,
+           const Tensor& gradOutput_,
+           const Tensor& input,
+           const Tensor& indices)
+{
+  TensorArg grad_input_arg{ gradInput, "gradInput", 1 };
+  TensorArg grad_output_arg{ gradOutput_, "gradOutput_", 2 };
+  TensorArg input_arg{ input, "input", 3 };
+  TensorArg indices_arg{ indices, "indices", 4 };
+
+  checkAllSameGPU("adaptive_max_pool3d_out_cuda",
+                 {grad_input_arg, grad_output_arg, input_arg, indices_arg});
+
+  const Tensor gradOutput = gradOutput_.contiguous();
+
+  gradInput.resize_as_(input);
+  gradInput.zero_();
+
+  int64_t sizeD, isizeT, isizeH, isizeW;
+  int64_t osizeT, osizeH, osizeW;
+  int64_t totalZ;
+
+  if (input.ndimension() == 4) {
+    sizeD = input.size(0);
+    isizeT = input.size(1);
+    isizeH = input.size(2);
+    isizeW = input.size(3);
+
+    osizeT = gradOutput.size(1);
+    osizeH = gradOutput.size(2);
+    osizeW = gradOutput.size(3);
+  } else {
+    sizeD = input.size(1);
+    isizeT = input.size(2);
+    isizeH = input.size(3);
+    isizeW = input.size(4);
+
+    osizeT = gradOutput.size(2);
+    osizeH = gradOutput.size(3);
+    osizeW = gradOutput.size(4);
+  }
+
+  bool atomic = (isizeW%osizeW != 0) || (isizeH%osizeH != 0) || (isizeT%osizeT != 0);
+
+  if (input.ndimension() == 4) {
+    totalZ = sizeD * osizeT;
+  } else {
+    int sizeB = input.size(0);
+    totalZ = sizeB * sizeD * osizeT;
+  }
+
+  int64_t offsetZ = 0;
+  dim3 threads(32, 8);
+  // each H*W plane is processed by blocksH thread blocks
+  int blocksH = std::max((int)(16L / totalZ), 1);
+  while (totalZ > 0) {
+    dim3 blocks(totalZ > 65535 ? 65535 : totalZ, blocksH);
+
+    if (atomic)
```

Comment:
Nittiest of nits: I probably would have just put the `atomic` conditional inside of the dynamic dispatch, rather than outside of it.